ui-v2/src/routes/runs/index.tsx

```diff
@@ -1,9 +1,199 @@
+		void context.queryClient.prefetchQuery(buildCountTaskRunsQuery());
+		void context.queryClient.prefetchQuery(
+			buildPaginateFlowRunsQuery(deps, 30_000),
+		);
+	},
+	wrapInSuspense: true,
 });
 
+const usePagination = () => {
+	const search = Route.useSearch();
+	const navigate = Route.useNavigate();
+
+	const pagination: PaginationState = useMemo(
+		() => ({
+			page: search.page ?? 1,
+			limit: search.limit ?? 10,
+		}),
+		[search.page, search.limit],
+	);
+
+	const onPaginationChange = useCallback(
+		(newPagination: PaginationState) => {
+			void navigate({
+				to: ".",
+				search: (prev) => ({
+					...prev,
+					page: newPagination.page,
+					limit: newPagination.limit,
+				}),
+				replace: true,
+			});
+		},
+		[navigate],
+	);
+
+	return [pagination, onPaginationChange] as const;
+};
+
+const useSort = () => {
+	const search = Route.useSearch();
+	const navigate = Route.useNavigate();
+
+	const onSortChange = useCallback(
+		(sort: SortFilters) => {
+			void navigate({
+				to: ".",
+				search: (prev) => ({ ...prev, sort, page: 1 }),
+				replace: true,
+			});
+		},
+		[navigate],
+	);
+
+	return [search.sort, onSortChange] as const;
+};
+
+const useHideSubflows = () => {
+	const search = Route.useSearch();
+	const navigate = Route.useNavigate();
+
+	const onHideSubflowsChange = useCallback(
+		(hideSubflows: boolean) => {
+			void navigate({
+				to: ".",
+				search: (prev) => ({ ...prev, "hide-subflows": hideSubflows, page: 1 }),
+				replace: true,
+			});
+		},
+		[navigate],
+	);
+
+	return [search["hide-subflows"], onHideSubflowsChange] as const;
+};
+
+const useTab = () => {
+	const search = Route.useSearch();
+	const navigate = Route.useNavigate();
+
+	const onTabChange = useCallback(
+		(tab: string) => {
+			void navigate({
+				to: ".",
+				search: (prev) => ({
+					...prev,
+					tab: tab as "flow-runs" | "task-runs",
+					page: 1,
+				}),
+				replace: true,
+			});
+		},
+		[navigate],
+	);
+
+	return [search.tab, onTabChange] as const;
+};
+
 function RouteComponent() {
-	return "🚧🚧 Pardon our dust! 🚧🚧";
+	const queryClient = useQueryClient();
+	const search = Route.useSearch();
+	const [pagination, onPaginationChange] = usePagination();
+	const [sort, onSortChange] = useSort();
+	const [hideSubflows, onHideSubflowsChange] = useHideSubflows();
+	const [tab, onTabChange] = useTab();
+
+	const [
+		{ data: flowRunsCount },
+		{ data: taskRunsCount },
+		{ data: flowRunsPage },
+	] = useSuspenseQueries({
+		queries: [
+			buildCountFlowRunsQuery(),
+			buildCountTaskRunsQuery(),
+			buildPaginateFlowRunsQuery(buildPaginationBody(search), 30_000),
+		],
+	});
+
+	const flowRuns = flowRunsPage?.results ?? [];
+
+	const onPrefetchPage = useCallback(
+		(page: number) => {
+			const filter = buildPaginationBody({
+				...search,
+				page,
+			});
+			void queryClient.prefetchQuery(
+				buildPaginateFlowRunsQuery(filter, 30_000),
+			);
+		},
+		[queryClient, search],
+	);
```

Comment:
To avoid suspending the page we need to ensure that ALL the data needed in the component tree is prefetched. In this case, that means we need to perform a call to the `/ui/flow_runs/count-task-runs` route for each retrieved flow run.